### PR TITLE
Issue 922: allow run bookie together with broker

### PR DESCRIPTION
**Pull request by zhaijack**
### Motivation

To simplify deployment, it will be good to run bookie along with broker. so people don't have to run separate bookkeeper cluster. It is good for small-size or midsize deployments

### Modifications

Add options in PulsarBrokerStarter to allow run bookie together with broker

### Result

PulsarBrokerStarter command's old behaviour will not change; user now could add new options to start bookie or bookie auto recovery together with broker.

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java [hunk 1]
```diff
@@ -56,12 +138,62 @@ public static void main(String[] args) throws Exception {
         String configFile = args[0];
         ServiceConfiguration config = loadConfig(configFile);
 
+        CommandLine cmdLine = readCommandLine(args);
+        boolean runBookie = runBookieTogether(cmdLine);
+        boolean runBookieAutoRecovery = runBookieAutoRecoveryTogether(cmdLine);
+        final BookieServer bookieServer;
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java]: I would suggest wrapping `BookieServer` `AutoRecoveryMain` and `StatsProvider` into a static inner class called `PulsarBookieStarter` and provide a method to `start` and a method to `stop`. so it is easier for unit testing. You can unit test `PulsarBookieStarter` as well.

File: pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java [hunk 2]
```diff
@@ -44,8 +55,79 @@ private static ServiceConfiguration loadConfig(String configFile) throws Excepti
         return config;
     }
 
+    private static final Options OPTS = new Options();
```
Comment (merlimat) [inline pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java]: We use JCommander for parsing CLI options in the other tools, eg: see https://github.com/apache/incubator-pulsar/blob/45ea0bc45e67751711708ecd7ba8f446ff34c67a/pulsar-broker/src/main/java/org/apache/pulsar/PulsarStandaloneStarter.java#L52

Comment (zhaijack) [reply pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java]: thanks, will change it.

File: pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java [hunk 3]
```diff
@@ -44,8 +55,79 @@ private static ServiceConfiguration loadConfig(String configFile) throws Excepti
...
+    static {
+        OPTS.addOption("b", "bookie", false, "Run Bookie Server together");
```
[the remark this answers is not in the document]
Comment (zhaijack) [reply pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java]: Thanks. will change it.

File: pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java [hunk 4]
```diff
@@ -44,8 +55,79 @@ private static ServiceConfiguration loadConfig(String configFile) throws Excepti
...
+            return false;
+        }
+    }
+
+    private static boolean runBookieAutoRecoveryTogether(CommandLine cmdLine) throws IllegalArgumentException {
+        if (cmdLine.hasOption('a') && cmdLine.hasOption('c')) {
+            return true;
+        } else if(cmdLine.hasOption('a')) {
+            printUsage();
+            throw new IllegalArgumentException("No configuration file for bookie auto recovery");
```
[the remark this answers is not in the document]
Comment (zhaijack) [reply pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java]: Thanks, will change it.

File: pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java [hunk 5]
```diff
@@ -44,8 +55,79 @@ private static ServiceConfiguration loadConfig(String configFile) throws Excepti
...
+
+        String bookieConfigFile = cmdLine.getOptionValue("c");
+        ServerConfiguration bookieConf = new ServerConfiguration();
+        try {
+            bookieConf.loadConf(new File(bookieConfigFile).toURI().toURL());
+            bookieConf.validate();
+            log.info("Using bookie configuration file {}", bookieConfigFile);
+        } catch (MalformedURLException e) {
+            log.error("Could not open configuration file: {}", bookieConfigFile, e);
+            throw new IllegalArgumentException();
```
[the remark this answers is not in the document]
Comment (zhaijack) [reply pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java]: Thanks, will change it.

File: pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java [hunk 6]
```diff
@@ -44,8 +55,79 @@ private static ServiceConfiguration loadConfig(String configFile) throws Excepti
...
+        } catch (ConfigurationException e) {
+            log.error("Malformed configuration file: {}", bookieConfigFile, e);
```
[the remark this answers is not in the document]
Comment (zhaijack) [reply pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java]: Thanks, will change it.

File: pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java (hunk 1 above)
Comment (zhaijack) [reply pulsar-broker/src/main/java/org/apache/pulsar/PulsarBrokerStarter.java]: Thanks, will change it.

File: pulsar-broker/src/test/java/org/apache/pulsar/PulsarBrokerStarterTest.java [hunk 7]
```diff
@@ -270,4 +277,70 @@ public void testMainWithNoArgument() throws Exception {
...
+     * Verifies that the main throws {@link IllegalArgumentException} when no config file for bookie is given.
+     */
+    @Test
+    public void testMainRunBookieEmptyConfig() throws Exception {
+        try {
+            File testConfigFile = createValidBrokerConfigFile();
+            String[] args = {testConfigFile.getAbsolutePath(), "-a", "-b", "-c", testConfigFile.getAbsolutePath()};
+            PulsarBrokerStarter.main(args);
+            Assert.fail("Effectively empty config file for bookie should've raised NoWritableLedgerDirException!");
+        } catch (LedgerDirsManager.NoWritableLedgerDirException e) {
```
[the remark this answers is not in the document]
Comment (zhaijack) [reply pulsar-broker/src/test/java/org/apache/pulsar/PulsarBrokerStarterTest.java]: thanks, will add a comments for it.

**Comment by zhaijack** [comment]
tested in docker to run with new command:
`PulsarBrokerStarter broker.conf -ra -rb -bc bookkeeper.conf`
seems working fine.

**Comment by zhaijack** [comment]
@merlimat, If we using config file way, we may also need reading the bookie-config-file from command-line. That means command line may have to change something, It is not purely through the broker config file.

Since we may have to change the cli, so changed to using purely cli options as it is now.

**Comment by merlimat** [comment]
@zhaijack Right now, when starting broker, we default to `conf/broker.conf` and the same happens for bookies with `conf/bookkeeper.conf`. These are defined in `bin/pulsar`: 

```shell
DEFAULT_BROKER_CONF=$PULSAR_HOME/conf/broker.conf
DEFAULT_BOOKKEEPER_CONF=$PULSAR_HOME/conf/bookkeeper.conf
```

 We could continue that way, and use the `BOOKKEEPER_CONF` variable to locate the file.

**Comment by zhaijack** [comment]
@merlimat , Thanks for the info.  Excuse me for not expressing well.  
Regarding an option in the broker config to run bookie together,   My concern is like this:

currently, command `PulsarBrokerStarter` is called as this:
``` 
$PulsarBrokerStarter $DEFAULT_BROKER_CONF
```

If we set an option runBookieTogether in `broker.conf`, the command to run bookie together would be like:
``` 
$PulsarBrokerStarter $DEFAULT_BROKER_CONF $DEFAULT_BOOKKEEPER_CONF
```
In this way, both the broker.conf and command - PulsarBrokerStarter usage need change. 
Since we may have to change the command(to add a new config file for bookie), so changed to using purely command line options maybe better.

**Comment by sijie** [comment]
@merlimat can you review @merlimat 's comment? let's see how can we move this forward.

**Comment by merlimat** [comment]
@zhaijack sorry for late reply, your argument is sound. 

I actually don't like a lot a the `PulsarBrokerStarter $DEFAULT_BROKER_CONF` already. (blame myself for picking that long time back...) but I'd prefer to also have an option for passing/changing the `$DEFAULT_BROKER_CONF`, like `--broker-conf $DEFAULT_BROKER_CONF` or defaulting to `./conf/broker.conf` if not specified. Same could be done for `bookkeeper.conf` argument. 

In any case it's a bit of a different matter. :)

**Pull request merged**